### PR TITLE
More improvements to the Controlled infrastructure

### PR DESCRIPTION
**Pull request by tanujkhattar**
Fixes https://github.com/quantumlib/Qualtran/issues/865
Fixes https://github.com/quantumlib/Qualtran/issues/686

The improvements are -

- `GateWithRegisters.controlled()` now returns `Controlled` bloq
- `cirq_gate_to_bloq` now maps a `cirq.Controlled` gates to `Controlled` bloq.
- `Controlled` bloq is now a `GateWithRegisters`. This is required because we want to support cirq style simulations for circuits containing `Controlled(Rx())` bloqs and `BloqAsCirqGate(Controlled())` is not sufficient because we need special logic to derive the unitary matrix for controlled bloqs (eg: when subbloq is a GateWithRegisters but doesn't have only THRU registers). 

After this PR, we should never need to invoke the `cirq.Controlled` gates in either cirq-style or bloq-style API in Qualtran and we should always use `Controlled` bloq in all cases. All existing uses of `cirq.Controlled` gates would be updated to `Controlled` bloq when invoking the bloq-style API using the `cirq_gate_to_bloq` mapping. 

This is the final PR in my series of PRs to improve the controlled infrastructure. cc @mpharrigan

### CODE REVIEW

File: qualtran/_infra/gate_with_registers.py [hunk 1]
```diff
@@ -311,23 +310,40 @@ def on_registers(
...
-
-        controlled_gate = cirq.ControlledGate(
-            self,
-            num_controls=num_controls,
-            control_values=control_values,
-            control_qid_shape=control_qid_shape,
-        )
-        ctrl_spec = CtrlSpec.from_cirq_cv(controlled_gate.control_values)
-        return BloqAsCirqGate(Controlled(self, ctrl_spec))
+    ) -> 'GateWithRegisters':
```
Comment (anurudhp) [inline qualtran/_infra/gate_with_registers.py:326]: Would it be useful to provide overloaded type hints?
```py
    # pylint: disable=arguments-renamed
    @overload
    def controlled(
        self,
        num_controls: Optional[int] = None,
        control_values=None,
        control_qid_shape: Optional[Tuple[int, ...]] = None,
    ) -> 'GateWithRegisters':
        ...

    # pylint: disable=signature-differs
    @overload
    def controlled(self, ctrl_spec: 'CtrlSpec') -> 'GateWithRegisters':
        ...

    # pylint: disable=arguments-renamed
    def controlled(self, num_controls=None, control_values=None, control_qid_shape=None):
```

Right now it's difficult to understand the usage without looking at the function body.

File: qualtran/_infra/adjoint.py [hunk 2]
```diff
@@ -150,6 +150,13 @@ def decompose_from_registers(
             return cirq.inverse(self.subbloq.decompose_from_registers(context=context, **quregs))
         return super().decompose_from_registers(context=context, **quregs)
 
+    def _circuit_diagram_info_(self, args: 'cirq.CircuitDiagramInfoArgs'):
```
Comment (mpharrigan) [inline qualtran/_infra/adjoint.py:153]: return type annotation

File: qualtran/_infra/controlled.py [hunk 3]
```diff
@@ -410,6 +411,18 @@ def add_my_tensors(
...
         tn.add(qtn.Tensor(data=data, inds=out_ind + in_ind, tags=[self.short_name(), tag]))
 
+    def _unitary_(self):
+        if isinstance(self.subbloq, GateWithRegisters):
+            # subbloq is a cirq gate, use the cirq-style API to derive a unitary.
+            return cirq.unitary(
+                cirq.ControlledGate(self.subbloq, control_values=self.ctrl_spec.to_cirq_cv())
+            )
+        if all(reg.side == Side.THRU for reg in self.subbloq.signature):
+            # subbloq has only THRU registers, so the unitary is well defined.
```
Comment (mpharrigan) [inline qualtran/_infra/controlled.py:421]: pycharm tells me this is supposed to be "well-defined"

File: qualtran/_infra/gate_with_registers.py (hunk 1 above)
Comment (mpharrigan) [reply qualtran/_infra/gate_with_registers.py:326]: yeah, I think this needs a big docstring that describes the two modes of operation

File: qualtran/bloqs/phase_estimation/text_book_qpe.py [hunk 4]
```diff
@@ -220,13 +222,14 @@ def decompose_from_registers(
...
+        target_qubits = merge_qubits(self.target_registers, **target_quregs)
 
-        phase_qubits = quregs['qpe_reg']
+        phase_qubits: Sequence[cirq.Qid] = quregs['qpe_reg'].flatten()
 
         yield self.ctrl_state_prep.on(*phase_qubits)
         for i, qbit in enumerate(phase_qubits[::-1]):
-            yield cirq.pow(unitary_op.controlled_by(qbit), 2**i)
+            yield Power(self.unitary.controlled(), 2**i).on(qbit, *target_qubits)
+            # yield cirq.pow(unitary_op.controlled_by(qbit), 2**i)
```
Comment (mpharrigan) [inline qualtran/bloqs/phase_estimation/text_book_qpe.py:232]: remove old code

**Review by mpharrigan — APPROVED** [review]
looks good. approval but please fix the issues raised

File: qualtran/bloqs/qsp/generalized_qsp.py [hunk 5]
```diff
@@ -382,14 +374,15 @@ def decompose_from_registers(
...
         for signal_rotation in self.signal_rotations[1:]:
             if num_inverse_applications > 0:
                 # apply C-U^\dagger
-                yield self.U.adjoint().on_registers(**quregs).controlled_by(signal_qubit)
+                yield self.U.adjoint().controlled().on(signal_qubit, *flat_qubits_for_u)
                 num_inverse_applications -= 1
             else:
                 # apply C[0]-U
-                yield self.U.on_registers(**quregs).controlled_by(signal_qubit, control_values=[0])
+                yield self.U.controlled(control_values=[0]).on(signal_qubit, *flat_qubits_for_u)
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/generalized_qsp.py:385]: Should we default to using qualtran style controls `CtrlSpec(cvs=0)` instead of cirq style `control_values=[0]`?

File: qualtran/_infra/gate_with_registers.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/_infra/gate_with_registers.py]: Added a detailed docstring and the overloads. Also added tests to verify the different cases raise errors.

File: qualtran/_infra/adjoint.py (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/_infra/adjoint.py]: done

File: qualtran/_infra/controlled.py (hunk 3 above)
Comment (tanujkhattar) [reply qualtran/_infra/controlled.py]: reworded

File: qualtran/bloqs/phase_estimation/text_book_qpe.py (hunk 4 above)
Comment (tanujkhattar) [reply qualtran/bloqs/phase_estimation/text_book_qpe.py]: Reverted the file so now there are no changes and the original implementation uses bloqs instead of cirq gates.

File: qualtran/bloqs/qsp/generalized_qsp.py (hunk 5 above)
Comment (tanujkhattar) [reply qualtran/bloqs/qsp/generalized_qsp.py]: Reverted this change because `self.U.on_registers(**quregs).controlled_by(signal_qubit)` now uses the `Controlled` bloq instead of cirq gate. The `op.controlled_by` still expects the cirq-style API since we don't have a custom `GWROperation` where we can override the `controlled_by` to expect the bloq-style API as well. But I think it's fine right now to do things this way

**Comment by tanujkhattar** [comment]
@mpharrigan I'm ready to merge this but feel free to take another look since I've made some minor upgrades. Using `gwr.controlled().on_registers()` and `gwr.on_registers().controlled_by` are now equivalent and both result `GateOperation(Controlled(gwr))`  where the wrapped gate is a the `Controlled` bloq. 

Tl;Dr - the discrepancy in usage from a users perspective that I described in Qualtran sync today is now gone and users don't need to worry about the intricacies when using the Cirq or Bloq style APIs.

File: qualtran/_infra/gate_with_registers.py [hunk 6]
```diff
@@ -311,39 +323,142 @@ def on_registers(
...
+        to generate a CtrlSpec from the cirq-style API and thus easily support both Cirq & Bloq
+        APIs. For example
+
+        >>> class CustomGWR(GateWithRegisters):
+        >>>     def controlled(self, *args, **kwargs) -> 'Bloq':
+        >>>         ctrl_spec = self._get_ctrl_spec(*args, **kwargs)
+        >>>         # Use ctrl_spec to construct a controlled version of `self`.
+
+        Args:
+            num_controls:
```
Comment (mpharrigan) [inline qualtran/_infra/gate_with_registers.py:359]: ...

**Comment by mpharrigan** [comment]
where is the logic for `controlled_by`?

**Comment by tanujkhattar** [comment]
> where is the logic for controlled_by?

No override is needed because of the way Cirq and Qualtran works. `GateOperation.controlled_by` delegates to `Gate.controlled()` by default, which now uses `GateWithRegisters.controlled()` which returns a `Controlled(GateWithRegisters)` which is a cirq gate and gets wrapped in the `GateOperation(controlled_bloq)`. 

This was true today morning as well, I just realized it when making these changes and thus added tests to `gate_with_registers_test.py` to make this concrete.

File: qualtran/_infra/gate_with_registers.py (hunk 6 above)
Comment (tanujkhattar) [reply qualtran/_infra/gate_with_registers.py]: Fixed

**Comment by tanujkhattar** [comment]
Merging this now
